docs: Emulation improvements (#17090)
microsoft/playwright: docs/src/emulation.md

title: "Emulation"
---

> 

Reviewer: Sketch:
- intro - maybe expand on the feature list?
- Devices in the config + library tab with `newContext()`. Example should have "Desktop Chromium" + "iPhone 11".
- Viewport - included in the device, but you can override for some tests with `test.use()` + library tab with `page.setViewportSize()`.
- Locale and timezone - can set globally in the config, and override for some tests with `test.use()`. Snippet with test.use + library tab with `newContext()`.
- Permissions - config snippet + library tab.
- Geolocation is like locale.
- Color scheme and media is like locale.
- User agent is included in the device. If you really need to test a different user agent string, test.use() + library tab.
- javaScriptEnabled to test how your page behaves with no javascript. Example with test.use + library tab.

Playwright allows overriding various parameters such as `viewportSize`, `deviceScaleFactor`, `locale`, `timezone`, `colorScheme`, `geolocation` and more.

## Devices
* langs: js, csharp, python

> Playwright comes with a registry of device parameters for selected mobile devices. It can be used to simulate browser behavior on a specific mobile device. All pages created in the context will share the same device parameters.

Reviewer: ```suggestion
Playwright comes with a registry of device parameters for selected mobile devices. It can be used to simulate browser behavior on a specific mobile device. All tests will run with the specified device parameters.
```

```js tab=js-ts
// playwright.config.ts
import { type PlaywrightTestConfig, devices } from '@playwright/test';

const config: PlaywrightTestConfig = {
  projects: [
    // "Pixel 4" tests use Chromium browser.
    {
      name: 'Pixel 4',
      use: {

>         browserName: 'chromium',

Reviewer: ```suggestion
```

        ...devices['Pixel 4'],
      },
    },
  ],
};
export default config;
```

```js tab=js-js
// playwright.config.js
// @ts-check
const { devices } = require('@playwright/test');

/** @type {import('@playwright/test').PlaywrightTestConfig} */
const config = {
  projects: [
    // "Pixel 4" tests use Chromium browser.
    {
      name: 'Pixel 4',
      use: {

>         browserName: 'chromium',

Reviewer: ```suggestion
```

        ...devices['Pixel 4'],
      },
    },
  ],
};

module.exports = config;
```

```js tab=js-library
const { chromium, devices } = require('playwright');
const browser = await chromium.launch();

> const pixel4 = devices['Pixel 4'];

Reviewer: Change other languages as well.

const context = await browser.newContext({

>   ...pixel2,

Reviewer: ```suggestion
  ...pixel4,
```

    TimezoneId = "Europe/Berlin"
});
```
## Permissions

> Emulate the permissions to show system notifications or geolocation with [`method: BrowserContext.grantPermissions`].

Reviewer: ```suggestion
Allow app to show system notifications.
```

)
```

> Allow test to show current location.

Reviewer: ```suggestion
Allow test to request current location.
```

await context.GrantPermissionsAsync(new[] { "geolocation" });
```

> Grant notifications access from a specific domain.

Reviewer: ```suggestion
Allow notifications for a specific domain.
```

```csharp
await context.ClearPermissionsAsync();
```
## Geolocation

> Create a test with `"geolocation"` permissions granted and geolocation set to a specific area with [`method: BrowserContext.setGeolocation`].

Reviewer: ```suggestion
Create a test with `"geolocation"` permissions granted and geolocation set to a specific area.
```

Change the location later:

> ```js

Reviewer: I think it was good like that. Otherwise, we lack the snippet for js. I'd only replace `context` with `page.context()` to start with the page that is in all the snippets.

Author: your right. moved it back
